src/mds/MDLog.cc

```diff
@@ -709,6 +742,7 @@ void MDLog::trim(int m)
 
   // discard expired segments and unlock submit_mutex
   _trim_expired_segments();
+  submit_mutex.lock();
```

Comment:
So, this is not OK in this function anymore, as I see you've agreed to remove the submit mutex lock at line 624. To follow that, you'd probably want to get rid of other submit_mutex operations in this function, and also call the public version of `trim_expired_segments()` which would take the lock and then call `_trim_expired_segments()` which would unlock. This will be transparent to this function and leave it with the submit mutex unheld as it is the case now when this function is entered.

Reply:
Oh dear -- this function is messy with multiple iterations to add lock and then remove them. Good catch, btw.